### PR TITLE
Image library adapter

### PR DESCRIPTION
**Pull request by Pepsi1x1**
This pull request convert DocxTemplater.Images into a library-agnostic base image package with adapter packages to extend the base with an implementation freeing from the potential license constraints or target framework constraints see in #110 and #16. It moves the target framework to netstandard2.1 for the core projects, a first step towards netstandard2.0. Add DocxTemplater.Images.ImageSharp as the default adapter package using the existing ImageSharp dependency to maintain compatibility with the published changes.

**Test infrastructure and coverage improvements:**

* Added a new test projects for the specific implementations of the Images project and test contracts in the main Images test project so that implementers don't need to duplicate test code to provide a new package.
* Implemented a comprehensive test class `BclImageMetadataReaderTest` that covers reading metadata for PNG, GIF, BMP, TIFF, and JPEG images, including edge cases such as malformed data, safe parsing limits, and error handling.

**Library enhancements:**

* Introduced a new `BCL` project that configures an `ImageFormatter` with a dependency-free BCL metadata reader, simplifying usage for consumers who want a pure BCL implementation.

**Project configuration changes:**

* Changed the `TargetFrameworks` property in `Directory.Build.props` from multiple .NET versions (`net8.0;net9.0;net10.0`) to `netstandard2.1`, potentially increasing compatibility with older .NET implementations.

### CODE REVIEW

File: DocxTemplater.Images.ImageSharp/DocxTemplater.Images.ImageSharp.csproj [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>net8.0;net9.0;net10.0</TargetFrameworks>
+    <GeneratePackageOnBuild>True</GeneratePackageOnBuild>
```
[the remark this answers is not in the document]
Comment (Pepsi1x1) [reply DocxTemplater.Images.ImageSharp/DocxTemplater.Images.ImageSharp.csproj:3]: Intentional as this maintains parity with the current implementation

File: DocxTemplater.Images/ImageFormatter.cs [hunk 2]
```diff
@@ -45,7 +72,7 @@ public void ApplyFormat(ITemplateProcessingContext templateContext, FormatterCon
                 }
 
             }
-            catch (Exception e) when (e is InvalidImageContentException or UnknownImageFormatException)
+            catch (ImageMetadataReadException e)
             {
                 throw new OpenXmlTemplateException("Could not detect image format", e);
             }
```
[the remark this answers is not in the document]
Comment (Pepsi1x1) [reply DocxTemplater.Images/ImageFormatter.cs:70]: Done

File: README.md [hunk 3]
```diff
@@ -54,8 +53,17 @@ Enhance DocxTemplater with these optional extension packages:
 | Package      | Description                             
 |--------------|-----------------------------------
 | [DocxTemplater.Images ](https://www.nuget.org/packages/DocxTemplater.Images)  |Enables embedding images in generated Word documents|
+| DocxTemplater.Images.Bcl (not published) | Proof-of-concept .NET BCL metadata adapter with no third-party image dependency |
 | [DocxTemplater.Markdown ](https://www.nuget.org/packages/DocxTemplater.Markdown)  | Allows use of Markdown syntax for generating parts of Word documents|
```
[the remark this answers is not in the document]
Comment (Pepsi1x1) [reply README.md:58]: Done

**Comment by Pepsi1x1** [comment]
Appreciate that you may not want to introduce the BCL package and take on that maintenance, the intent with including this was to prove the concept. Happy to remove that from the PR if you'd rather not take that part on. The idea here is that it would be possible to experiment with other providers like SkiaSharp or SyncFusion for example by publishing a specific package for this without breaking existing consumers and to make it simpler for consumers to bring their own provider with a smaller implementation.

**Comment by Amberg** [comment]
Thanks for this pull request! I've gone back and forth myself a few times about implementing the header parsing on our own to eliminate the dependencies, so I'm glad to see this. My take: the dependency-free version should be the default, and we can keep the ImageSharp library around as an optional adapter.

I'd like to take this in a slightly different direction so we land it without a breaking change, and split it into two independent steps rather than one combined PR. Right now two concerns are bundled together (the adapter refactor and a retarget to netstandard2.1), and I'd like to decouple them.

**Proposed direction**

**Step 1** — Drop the ImageSharp dependency, keep full backward compatibility (this PR)

The BCL reader has zero third-party dependencies, so it can move directly into the base DocxTemplater.Images package and become the built-in default:

- Move BclImageMetadataReader into DocxTemplater.Images and rename it to something neutral, e.g. DefaultImageMetadataReader / HeaderImageMetadataReader.
- Re-add the parameterless constructor so existing code keeps compiling:

```
public class ImageFormatter : IFormatter, IImageServiceProvider
{
    // Default: dependency-free header reader → no breaking change
    public ImageFormatter()
        : this(new DefaultImageMetadataReader())
    {
    }

    public ImageFormatter(IImageMetadataReader imageMetadataReader) { ... }
    public ImageFormatter(IImageService imageService) { ... }
}
```

- Keep the IImageMetadataReader / IImageService constructors for opt-in adapters.
- Drop the separate DocxTemplater.Images.Bcl PoC package (the metadata reader lives in the base package now).
- ImageSharp becomes an optional opt-in adapter for anyone who wants its more robust decoding: new ImageFormatter(new ImageSharpImageMetadataReader()).
- README: add a short migration note (default is now dependency-free; ImageSharp is opt-in).

Net result:
new ImageFormatter() works again (no breaking change), the base package is dependency-free, and issue #110 is addressed.

**Step 2** — netstandard / multi-targeting (separate PR)

Please pull the framework retarget out of this PR.
Once the ImageSharp dependency is gone, changing the target framework is a clean, isolated follow-up.

**Required changes for this PR**

1. Remove the committed test artifact DocxTemplater.Test/TestResults/full-suite.trx (~12.9 MB, 247k lines) — this would bloat the git history permanently. Please also add TestResults/ and *.trx to .gitignore.
2. Revert the Directory.Build.props retarget (move to Step 2 as above). This also removes most of the compat noise automatically.
3. Remove the unused helpers OpenXmlHelper.GetFirstChildWithTagName and GetFirstDescendantWithTagName — they aren't referenced anywhere.

**Minor (non-blocking)**

- ImageFormatter(IImageService): for a non-ImageService implementation the factory returns the same instance on every CreateImageService() call, so the mutable state (m_imagePartRelIdCache, m_currentRoot) would leak across runs. The concrete-ImageService path already guards against this — the fallback path doesn't.
- BCL JPEG MaxJpegMetadataBytesToScan counts skipped-segment payload bytes as well, not just bytes actually scanned, so a legitimate JPEG with a large ICC profile before the SOF could trip the limit. A short comment or test would be good.

Happy to help with any of this. Thanks again — with the dependency-free default I think this is a really nice improvement.

**Comment by Pepsi1x1** [comment]
Hi, totally agree with that direction and happy to go with that. The BCL implementation will need a little more attention to move from POC grade to production grade, as the additional comments you made allude to, such as hardening the scanning.

**Comment by Pepsi1x1** [comment]
@Amberg first pass is done, removing the netstandard parts from the PR and moving to BCL being the default implementation.

I'll do a hardening pass on the default implementation next so that it's not just a POC.

**Comment by Pepsi1x1** [comment]
Ready for review!

**Comment by Amberg** [comment]
This is looking really solid now — thanks for working through all the feedback so thoroughly.

A few small, non-blocking cleanups before merge:

1. Drop the leftover `IsExternalInit.cs` files in `DocxTemplate`" and `DocxTemplater.Images`. 
2. Rename the test project DocxTemplater.Images.Bcl.Test / BclImageMetadataReaderTest — it now exercises DefaultImageMetadataReader in the base DocxTemplater.Images package. Something like DocxTemplater.Images.Test / DefaultImageMetadataReaderTest would match the new reality, or fold it into DocxTemplater.Test.

Also, could you tidy up the commit history before merge? The branch adds the netstandard2.1 retarget and then reverts it again.

The ImageFormatter(IImageService) overload mostly exists to reject custom implementations at runtime and redundantly handle the concrete ImageService (which the IImageMetadataReader overload already covers). You could drop it in favor of IImageMetadataReader + Func<IImageService> so the type system doesn't advertise something it then throws on — but that's purely a taste call.

Thanks

**Comment by Pepsi1x1** [comment]
Ah yes, I missed those, I'll pull those out. 

For the test project I'll rename that to match the reality, as it seems like the right thing to do for new contributors to keep it as a separate assembly to show the testing seams for implementing a new adapter. 

I was considering the git history and wondered if a squash merge would cover that, if not I can have a look at replaying the commits. 

I'll take a look at the ctor again, as it really was intended to be a convenience but I think you're right at this point it's probably a "footgun".

**Comment by Amberg** [comment]
yes keep it simple, we can do a sqash merge

**Comment by Pepsi1x1** [comment]
Great, that should be all of the changes made.

File: DocxTemplater.Test/ImageFormatterTest.cs [hunk 4]
```diff
@@ -1,90 +1,72 @@
...
+
+    internal sealed class CoreImageFormatterContractTests : ImageFormatterContractTests
+    {
+        protected override DocxTemplater.Formatter.IFormatter CreateFormatter()
+        {
+            return new ImageFormatter(new CoreTestImageMetadataReader());
+        }
+    }
+
+    internal sealed class CoreTestImageMetadataReader : IImageMetadataReader
```
Comment (Amberg) [inline DocxTemplater.Test/ImageFormatterTest.cs]: Because there is now a functional reader in the base package we can use DefaultImageMetadataReader in Tests

File: DocxTemplater.Test/SvgImageFormatterTest.cs [hunk 5]
```diff
@@ -45,7 +45,7 @@ public void ProcessTemplateWithSvgImageInsertsImageCorrectly()
             memStream.Position = 0;
 
             var docTemplate = new DocxTemplate(memStream);
```
Comment (Amberg) [inline DocxTemplater.Test/SvgImageFormatterTest.cs:47]: With the DefaultImageMetadataReader this file and others don't change at all

**Review by Amberg — CHANGES_REQUESTED** [review]

File: DocxTemplater.Test/ImageFormatterTest.cs (hunk 4 above)
Comment (Pepsi1x1) [reply DocxTemplater.Test/ImageFormatterTest.cs]: Removed and updated

File: DocxTemplater.Test/SvgImageFormatterTest.cs (hunk 5 above)
Comment (Pepsi1x1) [reply DocxTemplater.Test/SvgImageFormatterTest.cs:47]: Good catch!

**Comment by Pepsi1x1** [comment]
I created some integration tests locally for my own sanity and noticed there may be a pre-existing bug in the ImageFormatter TransformSize method, the last if elseif else in that method, the else should be removed to support stretch correctly I think. I'll dig into that separately though

**Comment by Amberg** [comment]
> I created some integration tests locally for my own sanity and noticed there may be a pre-existing bug in the ImageFormatter TransformSize method, the last if elseif else in that method, the else should be removed to support stretch correctly I think. I'll dig into that separately though

Do you mean 

```
    // if both are set, the aspect ratio is kept
                var aspectRatio = (double)pixelWidth / pixelHeight;
                var newAspectRatio = (double)cxEmu / cyEmu;
                if (aspectRatio > newAspectRatio)
                {
                    cyEmu = (int)(cxEmu / aspectRatio);
                }
                else
                {
                    cxEmu = (int)(cyEmu * aspectRatio);
                }
```

The code keeps the aspect ratio. The image is not stretched to fill the full width and height. This is often the intended behavior.

**Comment by Pepsi1x1** [comment]
Yes that's the one, it's not touched by this PR, but one of the integration tests I'd written was for stretched width only or height only which then proportionally scales instead when it hits that code block. Not a blocker, just thought it worth some thought outside of the scope of this.

**Review by Amberg — APPROVED** [review]

**Comment by Pepsi1x1** [comment]
Thank you for your help on this, great to see my first contribution in and already in beta :)
